### PR TITLE
Update input validation for openssl_x509_certificate.

### PR DESCRIPTION
**Pull request by femiagbabiaka**
Fixes https://github.com/chef/chef/issues/14079.

## Description

TL;DR, as I understand it, you must have one of common_name or a subject_alt_name entry. If you have a common_name entry, it must match the subject_alt_name entry. But common_name is not required.

Currently, the resource implicitly requires common_name to be specified: https://github.com/chef/chef/blob/main/lib/chef/resource/openssl_x509_certificate.rb#L221. Implicit because common_name isn't marked as required: https://github.com/chef/chef/blob/main/lib/chef/resource/openssl_x509_certificate.rb#L101.

I believe the correct solution is to leave the common_name and subject_alt_name property definitions as-is, add an unless check on the nilness of common_name, and require that one of subject_alt_name or common_name be set.

## Related Issue

https://github.com/chef/chef/issues/14079
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [ ] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] If `Gemfile.lock` has changed, I have used `--conservative` to do it and included the full output in the Description above.
- [x] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

**Comment by femiagbabiaka** [comment]
I'm not 100% sure on how to resolve the last three bullet points in the checklist, and could use some assistance/advice in doing so. I assume that the documentation is generated by updating the property definition?

File: lib/chef/resource/openssl_x509_certificate.rb [hunk 1]
```diff
@@ -212,13 +212,18 @@ def request
         end
 
         def subject
+          if new_resource.common_name.nil? && new_resource.subject_alt_name.empty?
+            Chef::Log.fatal("Neither common_name nor subject_alt_name specified, one is required.")
+            raise "Neither common_name nor subject_alt_name specified, one is required."
```
Comment (tpowell-progress) [inline lib/chef/resource/openssl_x509_certificate.rb:217]: While this may be the correct way of doing things, I have concerns that this could break people by surprise. Can we do a deprecation first?

Comment (femiagbabiaka) [reply lib/chef/resource/openssl_x509_certificate.rb:217]: @tpowell-progress that makes sense, thanks. I'm new to deprecation warnings in chef resources, is this a good pattern to follow? https://github.com/chef/chef/blob/d06dc07c825f516752b9b26e443e9c6da38a9435/lib/chef/resource/archive_file.rb#L137

Comment (tpowell-progress) [reply lib/chef/resource/openssl_x509_certificate.rb:217]: Looking around, I think that pattern makes sense, provided that you can accomplish the deprecation warning without getting too messy. If for some reason affixing the deprecation warning doesn't look clean, we can dig deeper, but I'd recommend just pushing a reasonable try at it and we can see if makes sense.

**Review by jaymzh — CHANGES_REQUESTED** [review]
Requesting changes formally so that @femiagbabiaka sees the request.

**Review by jaymzh — APPROVED** [review]
Brilliant, thanks!

**Review by jaymzh — CHANGES_REQUESTED** [review]
Oh, you need DCO on all commits.

**Comment by femiagbabiaka** [comment]
@jaymzh thanks, missed that!

**Review by jaymzh — APPROVED** [review]
will merge after tests pass

**Comment by femiagbabiaka** [comment]
going to work on running these locally to tease out more/better error messages

**Comment by jaymzh** [comment]
I'd start by rebasing, honestly - your based on a super old commit which probably has many build fixes since then.

**Comment by jaymzh** [comment]
This license tests aren't your fault. https://github.com/chef/chef/pull/14663 should fix it when it gets in.

**Comment by jaymzh** [comment]
@femiagbabiaka  - we have fixed all the builds on main, can you rebase? So sorry for the delay!

**Comment by femiagbabiaka** [comment]
@jaymzh no problem, thanks!

**Comment by jaymzh** [comment]
The windows failure is not your fault, I'll get someone at Progress to restart the windows test there - the workers keep dying and they're working on that.

The skipped tests are tests for which the runners recently fell off the list of available actions runners, we'll get those nuked, but you can ignore them.
